### PR TITLE
Baseline solution offset

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
Since we moved to rank by surplus baseline is bleeding money due to bad gas estimates for its solutions which leads to collecting not enough fees.
These missing fees eventually have to be paid out of the solver's pocket.

# Changes
Add a  config parameter that allows increasing or decreasing a solution's gas estimate. According to @harisang this should be enough to stop the bleeding.
Obviously this is very hacky and we should a better solution soon which very likely involves simulating the solution.

### CODE REVIEW

File: crates/solvers/src/domain/eth/mod.rs [hunk 1]
```diff
@@ -57,6 +73,18 @@ impl std::ops::Add for Gas {
     }
 }
 
+impl std::ops::Add<SignedGas> for Gas {
+    type Output = Self;
+
+    fn add(self, rhs: SignedGas) -> Self::Output {
```
Comment (Mateo-mro) [inline crates/solvers/src/domain/eth/mod.rs:79]: We should also consider the sign of the `self`, what if `self` is also negative?

**Review by Mateo-mro — APPROVED** [review]

**Comment by squadgazzz** [comment]
How to properly decide on the config values? Didn't quite get it from the PR.

File: crates/solvers/src/domain/solver/baseline.rs [hunk 2]
```diff
@@ -153,9 +159,9 @@ impl Inner {
                     output.amount = cmp::min(output.amount, order.buy.amount);
                 }
 
+                let gas = route.gas() + self.solution_gas_offset;
```
Comment (fleupold) [inline crates/solvers/src/domain/solver/baseline.rs:162]: We already add the settlement gas overhead inside `Single::IntoSettlement` when computing the actual fee:

https://github.com/cowprotocol/services/blob/ac6580a2870099afc150c8e1527f1977dd659c74/crates/solvers/src/domain/solution.rs#L170-L177

Wouldn't this be the better place to add this? Maybe we could replace the settlement overhead with this parameter and increase it by whatever value @harisang suggested.

**Review by fleupold — COMMENTED** [review]
I think the "proper" fix would be to compute the expected revert risk cost based on the risk parameters and add that to the surplus fee we charge, but ok this seems simpler. If possible let's factor out the `SETTLEMENT_OVERHEAD` into this config so we don't introduce more random summands.

File: crates/solvers/src/domain/solver/baseline.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/solvers/src/domain/solver/baseline.rs:162]: Yeah, good idea. Conceptually the gas overhead should be a constant (as it should not depend on the interactions we do) so I didn't think about touching that. But that constant is obviously not good enough to make all estimators accurate so we might as well ditch it completely and only have a single configurable gas offset (possibly with the current constant value as the default).

**Comment by harisang** [comment]
> I think the "proper" fix would be to compute the expected revert risk cost based on the risk parameters and add that to the surplus fee we charge, but ok this seems simpler
As a comment here, this is not only about revert risk. For some reason we underestimate costs, and even if just restrict ourselves to successful settlements (i.e., not worry about penalties) and look at what Baseline settled, it is clear it underestimated fees in most settlements.

Here is a query i use that reveals this.
https://dune.com/queries/3431685

If you look at total results or per settlement results for Baseline, ,you will notice this.

In order to get some heuristic constants to correct gas estimations, I used this query
https://dune.com/queries/3558412?StartTime_d83555=2024-03-19+12%3A00%3A00&EndTime_d83555=2024-03-27+00%3A00%3A00&offset_n26d66=60000

Here, the offset is added on a per order basis and the winner is also adjusted. Specifically, we look at the wins of Baseline, we adapt the fees, and then check if it is still the winner, and only if it still is the winner, we check the updated cost coverage.

**Review by sunce86 — APPROVED** [review]

**Pull request merged**